introduced option to specify file encoding

d-chris opened the pull request:
fixes #283  

introduced option to specify file encoding when overwriting files.

## default values

- ini-file: `None`
- cli-option: `"utf-8"`

> ini-file always overwrite command line option

## command line option
```
--text-file-encoding=TEXT_FILE_ENCODING
                        Specify encoding for files.
```

## configuration file

```
  text_file_encoding (string):
                        Default encoding for text files.
```

pllim commented on pytest_doctestplus/plugin.py:
```diff
@@ -912,13 +919,13 @@ def test_filter(test):
 
 
-def write_modified_file(fname, new_fname, changes):
+def write_modified_file(fname, new_fname, changes, encoding=None):
```
Should the default here be consistent with addoption default (utf-8)?

bsipocz replied:
Yes, it should be consistent, though I'm not sure if we should change the default to utf-8 or leave as is now?

d-chris replied:
I decided to set `encoding=None` to avoid changing the behavior of the `write_modified_file` function since I am not familiar with your codebase. I wanted to ensure that the function behaves as it did before.

From a developer's perspective, I would prefer using  `"utf-8"` as the default encoding.

d-chris replied:
any changes required?

i guess the default value of  `write_modified_file` is uncritical.

default value for cli option defaults to "utf-8" which will be passed on.

pllim reviewed (commented):
Thanks!

More eyes on review would be good. The diff looks reasonable to me.

bsipocz commented:
@pllim - bugfix or enhancement?

bsipocz reviewed (changes requested):
This will definitely need a test, the one in the bug report issue would be great. And will eventually need an entry in the changelog, though we need to agree on if this is a feature or a bugfix first.

pllim commented:
> bugfix or enhancement?

Hmm... It does introduce a new option, so more like enhancement?

d-chris commented:
i tried to implement some tests

but seems with with `testdir.inline_run(p, "--doctest-plus-generate-diff=diff", "--text-file-encoding", encoding)` that the `DebugRunnerPlus` has an issue.
```
Traceback (most recent call last):
  File "C:\Python313\Lib\doctest.py", line 1395, in __run
    exec(compile(example.source, filename, "single",
         ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 compileflags, True), test.globs)
                 ^^^^^^^^^^^^^^^^^^^
  File "<doctest test_file_encoding_utf8.f[0]>", line 1
    print(☆)
          ^
SyntaxError: invalid character '☆' (U+2606)`
```

see commit f91f59704b974cc4b91d2542dd1b9798dd49c064

d-chris commented:
changed BR to draft

## todos

- fix issue with syntax error
- refactore my pytest functions, split into seperate tests for `diff` and `overwrite`
- write a small contribution.md / requirements-dev.txt, how i set up .venv to develop for windows

d-chris commented:
for testing implemented alternative fixture to generate temporary python files, due issue #285

d-chris commented:
somehow a can't trigger action within my fork,  any tips?

i created a new action (outside of the PR scope) -> pass https://github.com/d-chris/pytest-doctestplus/actions/runs/14041267921

so the small fix from 09979a3 solved the issue

d-chris reviewed (commented):
fixed in 09979a3

pllim commented:
Github Actions has to be enabled explicitly by maintainer if it detects a first time contributor. It is how they attempt to prevent bitcoin miners. Nothing you could have done on your side. I just enabled the CI. Thanks for your patience!

d-chris commented:
Thanks for your patience—this is my first GH contribution ever!

Is there anything else I need to do to finish the PR?

pllim commented:
Since @bsipocz requested changes, would be nice if she could re-review. 🙏

d-chris commented on pytest_doctestplus/plugin.py:
```diff
@@ -953,6 +960,9 @@ def pytest_terminal_summary(terminalreporter, exitstatus, config):
         return  # we do not report or apply diffs
+    
+    # get encoding to open file default ini=None or option="utf-8"
+    encoding = config.getini("text_file_encoding") or config.getoption("text_file_encoding")
```
any specification what has the higher priority cli-option or ini-file?

bsipocz replied:
This seems to be more consistent with the rest of the plugin, but I have to realise we're not 100% consistent 😱

d-chris replied:
should be solved with 4a639d7

bsipocz commented:
> Since @bsipocz requested changes, would be nice if she could re-review

Yeap, I plan to come back to this and the other PR later this week.

bsipocz commented on pytest_doctestplus/plugin.py:
On L355 we already use `doctest_encoding`. I wonder if we really need to have a different `text-file-encoding` or could reuse the already existing one?

d-chris replied:
I scrolled through the `pytest_addoption` options but didn't find anything suitable. (i did not check all 1k+ lines 😄)

I guess the [doctest_encoding](https://github.com/scientific-python/pytest-doctestplus/blob/cc0144a0a4161d2f988e8627c75bca366bc79b8f/pytest_doctestplus/plugin.py#L362C44-L362C60) option could be used, but it should also be made publicly available and documented for the `pytest --help` output.

i let the maintainers decide on this topic.

bsipocz replied:
Yes, because it's a standard pytest/doctest ini option, so we don't need to parse for it in doctestplus (no cli option for it) (https://docs.pytest.org/en/latest/how-to/doctest.html#encoding; but is also listed in ``pytest --help``).

So I would still go for using it, unless can show a specific use case where the cli option should be able to be different than what's stated in the ini.

d-chris replied:
@bsipocz 

i tried to change it to `doctest_encoding` but i run in an issue. this is only available as ini-setting.

|solution|con|
|-|-|
|introduce `doctest_encoding` as cli-option?| i guess users will be confused. because cli option will only work in doctest-plus|
| for cli option stick with `text-file-encoding`? | more confusing for ini  and cli two different parameters|
|stick with the current naming| user has to specify in the worst case two parameter|

bsipocz replied:
I would go with having this only as an ini option. Users can be specific in the config files, after all this is not something you want to turn on and off depending on the test run but something stays specific to the given package.

bsipocz reviewed (commented):
OK, this looks good to me. My only question is if we really need to add a new config option or reusing  the existing `doctest_encoding` would be sufficient. Have you looked into that option? Obviously it only affects the adding a new parser option and not the actual code fixes where you added the usage of encoding.

bsipocz commented:
Hmm, failures look relevant for pytest < 7.4, but I haven't looked into the details.

Possible options:
 - find an easy workaround that makes this feature work; or only have it for 7.4+
 - we can consider dropping support for older versions. We don't really follow SPEC0 for these infrastructure tools but keep support until we run into a blockage, and I would be open to discuss if this is a blockage (7.3 will just about pass the 2 year limit in a couple of days)

d-chris commented:
i use option `--config-file` to specify the ini file while running my tests.

whats confuses me i always run tox before a review, and i always pass all tests

bsipocz commented:
ahh, indeeed. `--config-file` was new in 7.4: https://github.com/pytest-dev/pytest/releases/tag/7.4.0

d-chris commented:
`-c` should do the trick 

```
pytest --version & pytest --help | findstr "config"
This is pytest version 5.1.0, imported from D:\temp\pytest5\.venv\lib\site-packages\pytest.py
                        configuration file raise errors.
  -c file               load configuration from `file` instead of trying to
                        locate one of the implicit configuration files.
test session debugging and configuration:
  -h, --help            show help message and configuration info
  --trace-config        trace considerations of conftest.py files.
```

bsipocz reviewed (approved):
This all looks good now. Thank you @d-chris for your contribution and patience during the review process.
